tiledb/sm/query/readers/ordered_dim_label_reader.h

```diff
@@ -0,0 +1,533 @@
+  bool incomplete() const;
+
+  /** Returns the status details reason. */
+  QueryStatusDetailsReason status_incomplete_reason() const;
+
+  /** Initialize the memory budget variables. */
+  Status initialize_memory_budget();
+
+  /** Performs a read query using its set members. */
+  Status dowork();
+
+  /** Resets the reader object. */
+  void reset();
+
+ private:
+  /* ********************************* */
+  /*         PRIVATE ATTRIBUTES        */
+  /* ********************************* */
+
+  /** UID of the logger instance */
+  inline static std::atomic<uint64_t> logger_id_ = 0;
+
+  /** Ranges to be retreived for the attribute. */
+  std::vector<Range> ranges_;
+
+  /** Stores the label attribute name. */
+  std::string label_name_;
+
+  /** Stores the dimension label type. */
+  Datatype label_type_;
+
+  /** Stores if the dimension label is var sized. */
+  bool label_var_size_;
+
+  /** Are the labels increasing? */
+  bool increasing_labels_;
+
+  /** Stores a pointer to the index dimension. */
+  const Dimension* index_dim_;
+
+  /** Save pointers to the non empty domain for each fragments. */
+  std::vector<const void*> non_empty_domains_;
+
+  /** Non empty domain for the array. */
+  Range non_empty_domain_;
+
+  /** Minimum tile index in the full domain. */
+  uint64_t tile_idx_min_;
+
+  /** Maximum tile index in the full domain. */
+  uint64_t tile_idx_max_;
+
+  /**
+   * Per fragment map to hold the result tiles. The key for the maps is the
+   * tile index.
+   */
+  std::vector<std::unordered_map<uint64_t, OrderedDimLabelResultTile>>
+      result_tiles_;
+
+  /** Tile index (inside of the domain) of the first tile of each fragments.
+   */
+  std::vector<uint64_t> domain_tile_idx_;
+
+  /**
+   * Stores the tile indexes (min/max) that can potentially contain the label
+   * value for each range start/end.
+   */
+  std::vector<RangeTileIndexes> range_tile_indexes_;
+
+  /** Total memory budget. */
+  uint64_t memory_budget_;
+
+  /* ********************************* */
+  /*           PRIVATE METHODS         */
+  /* ********************************* */
+
+  /** Performs a read on an ordered label array. */
+  void label_read();
+
+  /**
+   * Performs a read on an ordered label array.
+   *
+   * @tparam The index type.
+   */
+  template <typename IndexType>
+  void label_read();
+
+  /**
+   * Compute the tile index (inside of the full domain) of the first tile of
+   * each fragments.
+   *
+   * @tparam The index type.
+   */
+  template <typename IndexType>
+  void compute_domain_tile_indexes();
+
+  /**
+   * Compute the tile indexes (min/max) that can potentially contain the label
+   * value for each range start/end.
+   *
+   * @tparam The index type.
+   */
+  template <typename IndexType>
+  void compute_range_tile_indexes();
+
+  /**
+   * Compute the non empty domain for the index dimension.
+   *
+   * @tparam The index type.
+   */
+  template <typename IndexType>
+  void compute_non_empty_domain();
```

Comment:
Nit: Mention that this also computes the tile index location of the non-empty domain for the whole array.